File: lintcode/Clone_Graph.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
        // write your code here
        if(!node) return nullptr;
        
        UndirectedGraphNode *cGraph = new UndirectedGraphNode(node->label);
        unordered_map<UndirectedGraphNode *, UndirectedGraphNode *> map;
        map[node] = cGraph;
        
        stack<UndirectedGraphNode *> st;
        st.push(node);
        
        while(!st.empty()){
            UndirectedGraphNode *t = st.top();
            st.pop();
            
            for(int i = 0; i < (t->neighbors).size(); i++){
                UndirectedGraphNode *p = t->neighbors[i];
                if(map.find(p) == map.end()){
                    // copy nodes
                    UndirectedGraphNode *q = new UndirectedGraphNode(p->label);
                    map[p] = q;
                    st.push(p);
                }
                
                // copy edges
                map[t]->neighbors.push_back(map[p]);
            }    
        }
        
        return cGraph;
        
    }
};

File: lintcode/java/CloneGraph.java
/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     ArrayList<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
    /**
     * @param node: A undirected graph node
     * @return: A undirected graph node
     */
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode root) {
        if(root == null) return null;
        
        Map<UndirectedGraphNode, UndirectedGraphNode> nodeMap = new HashMap<>();
        Deque<UndirectedGraphNode> st = new ArrayDeque<>();
        st.push(root);
        UndirectedGraphNode nGraph = new UndirectedGraphNode(root.label);
        nodeMap.put(root, nGraph);
        
        while(!st.isEmpty()){
            UndirectedGraphNode t = st.pop();
            for(UndirectedGraphNode n : t.neighbors){
                if(!nodeMap.containsKey(n)){
                    UndirectedGraphNode nn = new UndirectedGraphNode(n.label);
                    // n node has not been explored yet
                    st.push(n);
                    nodeMap.put(n, nn);
                }
                //edge copy
                nodeMap.get(t).neighbors.add(nodeMap.get(n));
            }
        }
        
        return nGraph;
    }
}
